lib/propolis/src/vmm/hdl.rs

```diff
@@ -404,21 +414,148 @@ impl VmmHdl {
     }
 
+    // TODO: Hacky and strictly tied to migration versioning. Want to think of
+    // a better way to structure this.
+    pub fn export_vm(
+        &self,
+    ) -> std::result::Result<BhyveVmV1, MigrateStateError> {
+        Ok(BhyveVmV1::read(self)?)
+    }
+
     /// Export the global VMM state.
     pub fn export(
         &self,
     ) -> std::result::Result<Box<dyn Serialize>, MigrateStateError> {
-        Ok(Box::new(migrate::BhyveVmV1::read(self)?))
+        Ok(Box::new(BhyveVmV1::read(self)?))
     }
 
     /// Restore previously exported global VMM state.
     pub fn import(
         &self,
         deserializer: &mut dyn Deserializer,
     ) -> std::result::Result<(), MigrateStateError> {
-        let deserialized: migrate::BhyveVmV1 =
-            erased_serde::deserialize(deserializer)?;
-        deserialized.write(self)?;
+        let mut imported: BhyveVmV1 = erased_serde::deserialize(deserializer)?;
+
+        // Update guest timing-related data to adjust for migration time and
+        // movement across hosts before writing state back to the VMM.
+        self.adjust_timing_data(&mut imported.timing_info)?;
+
+        imported.write(self)?;
+        Ok(())
+    }
+
+    /// Update guest timing-related data to account for how much time has
+    /// elapsed since the VMM timing data was read on the source.
+    ///
+    ///
+    /// We need to update the guest-related fields of
+    /// [`BhyveVmV1::TimingInfoV1`] as such:
+    ///
+    /// - (`guest_tsc`)[TimingInfoV1::guest_tsc], the guest TSC
+    /// value at migration time, which is used by bhyve to calculate the guest's
+    /// TSC offset.
+    /// - (`boot_hrtime`)[TimingInfoV1::boot_hrtime], which represents
+    /// the hrtime on the host for when the VM booted. This is a relative value
+    /// used to normalize device timers on the kernel side, so it will be
+    /// negative if the guest has a longer uptime than the host it's running on.
+    ///
+    /// Note that this adjustment won't capture all the time that has passed
+    /// between reading data on the source and it being updated on the kernel
+    /// side; bhyve makes some additional similar adjustments to both of these
+    /// values to account for the latency from when we import the state to it
+    /// actually arriving in-kernel.
+    ///
+    /// In order to make those adjustments, we need to also update the
+    /// host-related fields of [TimingInfoV1]:
+    ///
+    /// - (`hrtime`)[TimingInfoV1::hrtime]: the host hrtime (monotonic clock)
+    /// when these adjustments were made
+    /// - (`hrestime`)[TimingInfoV1::hrestime]: the host hrestime (wall clock)
+    /// when these adjustments were made
+    ///
+    /// See the [`time_adjust`] module for more details about how the adjustment
+    /// calculations here are performed.
+    fn adjust_timing_data(
+        &self,
+        timing_data: &mut TimingInfoV1,
+    ) -> std::result::Result<(), MigrateStateError> {
```

Comment:
Might make it easier to reason about this function (and/or instrument it with logging) if it took a TimingInfoV1 argument, and emitted one through its Result type, rather than a mutable reference.  It'd be a pain to do in C (returning structs and all), but it seems more rust-like in this context. *shrug*

Reply:
done (`803081`)